controllers/apps/transformer_cluster_credential.go

```diff
@@ -140,62 +132,216 @@ func (t *clusterCredentialTransformer) buildClusterCredential(transCtx *clusterT
+	services []appsv1alpha1.Service, svcEndpoint *appsv1alpha1.ConnectionServiceEndpoint) (string, []corev1.ServicePort) {
+	return lookupMatchedService(services, svcEndpoint, func(svcName string) string {
+		return constant.GenerateComponentServiceName(clusterName, compName, svcName)
+	})
+}
+
+func lookupMatchedService(services []appsv1alpha1.Service, svcEndpoint *appsv1alpha1.ConnectionServiceEndpoint,
+	svcNameBuilder func(svcName string) string) (string, []corev1.ServicePort) {
+	for i, svc := range services {
+		if svc.Name == svcEndpoint.ServiceName {
+			return svcNameBuilder(svc.ServiceName), services[i].Spec.Ports
+		}
+	}
 	return "", nil
 }
 
-func buildCredentialEndpointFromService(credential appsv1alpha1.ConnectionCredential,
-	serviceName string, ports []corev1.ServicePort, data *map[string][]byte) {
-	port := int32(0)
-	if len(credential.PortName) == 0 {
-		port = ports[0].Port
-	} else {
-		for _, servicePort := range ports {
-			if servicePort.Name == credential.PortName {
-				port = servicePort.Port
+func buildPodEndpoint(namespace, clusterName, compName string, replicas int, compDef *appsv1alpha1.ComponentDefinition,
+	credential appsv1alpha1.ConnectionCredential) ([]string, string, error) {
+	var (
+		hosts = make([]string, 0)
+		port  string
+	)
+	if len(compDef.Spec.Runtime.Containers) == 0 {
+		return nil, "", fmt.Errorf("")
+	}
+
+	for ordinal := 0; ordinal < replicas; ordinal++ {
+		hosts = append(hosts, constant.GeneratePodFQDN(namespace, clusterName, compName, ordinal))
+	}
+
+	svcEndpoint := credential.Endpoint.PodEndpoint
+	container := compDef.Spec.Runtime.Containers[0]
+	if len(svcEndpoint.Container) > 0 {
+		for i, c := range compDef.Spec.Runtime.Containers {
+			if c.Name == svcEndpoint.Container {
+				container = compDef.Spec.Runtime.Containers[i]
+				break
+			}
+		}
+	}
+	if len(container.Ports) == 0 {
+		return nil, "", fmt.Errorf("")
+	}
+	if len(svcEndpoint.PortName) == 0 && len(container.Ports) > 1 {
+		return nil, "", fmt.Errorf("")
+	}
+	containerPort := container.Ports[0]
+	if len(svcEndpoint.PortName) > 0 {
+		for i, p := range container.Ports {
+			if p.Name == svcEndpoint.PortName {
+				containerPort = container.Ports[i]
 				break
 			}
 		}
 	}
+	if containerPort.HostPort > 0 {
+		port = fmt.Sprintf("%d", containerPort.HostPort)
+	} else {
+		port = fmt.Sprintf("%d", containerPort.ContainerPort)
+	}
+
+	return hosts, port, nil
+}
+
+func buildCredentialEndpointFromService(credential appsv1alpha1.ConnectionCredential, hosts []string, port string, data *map[string][]byte) {
 	// TODO(component): define the service and port pattern
-	(*data)["endpoint"] = []byte(fmt.Sprintf("%s:%d", serviceName, port))
-	(*data)["host"] = []byte(serviceName)
-	(*data)["port"] = []byte(fmt.Sprintf("%d", port))
+	endpoints := make([]string, 0)
+	if len(port) == 0 {
+		endpoints = hosts
+	} else {
+		for i := range hosts {
+			endpoints = append(endpoints, fmt.Sprintf("%s:%s", hosts[i], port))
```

Comment:
The format of the endpoint is host:port{separator}host:port. Should this be explained in the API?